// next/dist/data-client/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "useQuery", {
    enumerable: true,
    get: function() {
        return _reactQuery.useQuery;
    }
});
Object.defineProperty(exports, "usePaginatedQuery", {
    enumerable: true,
    get: function() {
        return _reactQuery.usePaginatedQuery;
    }
});
Object.defineProperty(exports, "useInfiniteQuery", {
    enumerable: true,
    get: function() {
        return _reactQuery.useInfiniteQuery;
    }
});
Object.defineProperty(exports, "useMutation", {
    enumerable: true,
    get: function() {
        return _reactQuery.useMutation;
    }
});
Object.defineProperty(exports, "MutateFunction", {
    enumerable: true,
    get: function() {
        return _reactQuery.MutateFunction;
    }
});
Object.defineProperty(exports, "queryClient", {
    enumerable: true,
    get: function() {
        return _reactQueryUtils.queryClient;
    }
});
Object.defineProperty(exports, "getQueryKey", {
    enumerable: true,
    get: function() {
        return _reactQueryUtils.getQueryKey;
    }
});
Object.defineProperty(exports, "getInfiniteQueryKey", {
    enumerable: true,
    get: function() {
        return _reactQueryUtils.getInfiniteQueryKey;
    }
});
Object.defineProperty(exports, "invalidateQuery", {
    enumerable: true,
    get: function() {
        return _reactQueryUtils.invalidateQuery;
    }
});
Object.defineProperty(exports, "setQueryData", {
    enumerable: true,
    get: function() {
        return _reactQueryUtils.setQueryData;
    }
});
Object.defineProperty(exports, "useQueryErrorResetBoundary", {
    enumerable: true,
    get: function() {
        return _reactQuery1.useQueryErrorResetBoundary;
    }
});
Object.defineProperty(exports, "QueryClient", {
    enumerable: true,
    get: function() {
        return _reactQuery1.QueryClient;
    }
});
Object.defineProperty(exports, "dehydrate", {
    enumerable: true,
    get: function() {
        return _hydration.dehydrate;
    }
});
Object.defineProperty(exports, "invoke", {
    enumerable: true,
    get: function() {
        return _invoke.invoke;
    }
});
var _exportNames = {
};
var _constants = _interopRequireWildcard(require("./constants"));
var _rpc = _interopRequireWildcard(require("./rpc"));
var _auth = _interopRequireWildcard(require("./auth"));
var _reactQuery = require("./react-query");
var _reactQueryUtils = require("./react-query-utils");
var _reactQuery1 = require("react-query");
var _hydration = require("react-query/hydration");
var _invoke = require("./invoke");
function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
        return obj;
    } else {
        var newObj = {
        };
        if (obj != null) {
            for(var key in obj){
                if (Object.prototype.hasOwnProperty.call(obj, key)) {
                    var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {
                    };
                    if (desc.get || desc.set) {
                        Object.defineProperty(newObj, key, desc);
                    } else {
                        newObj[key] = obj[key];
                    }
                }
            }
        }
        newObj.default = obj;
        return newObj;
    }
}
Object.keys(_constants).forEach(function(key) {
    if (key === "default" || key === "__esModule") return;
    if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
    if (key in exports && exports[key] === _constants[key]) return;
    Object.defineProperty(exports, key, {
        enumerable: true,
        get: function() {
            return _constants[key];
        }
    });
});
Object.keys(_rpc).forEach(function(key) {
    if (key === "default" || key === "__esModule") return;
    if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
    if (key in exports && exports[key] === _rpc[key]) return;
    Object.defineProperty(exports, key, {
        enumerable: true,
        get: function() {
            return _rpc[key];
        }
    });
});
Object.keys(_auth).forEach(function(key) {
    if (key === "default" || key === "__esModule") return;
    if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
    if (key in exports && exports[key] === _auth[key]) return;
    Object.defineProperty(exports, key, {
        enumerable: true,
        get: function() {
            return _auth[key];
        }
    });
});
