// next/dist/server/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.isBlockedPage = isBlockedPage;
exports.cleanAmpPath = cleanAmpPath;
exports.resultFromChunks = resultFromChunks;
exports.resultToChunks = resultToChunks;
exports.fixNodeFileTrace = exports.isInternalBlitzMonorepoDevelopment = exports.withFixNodeFileTrace = void 0;
var _constants = require("../shared/lib/constants");
function isBlockedPage(pathname) {
    return _constants.BLOCKED_PAGES.includes(pathname);
}
function cleanAmpPath(pathname) {
    if (pathname.match(/\?amp=(y|yes|true|1)/)) {
        pathname = pathname.replace(/\?amp=(y|yes|true|1)&?/, '?');
    }
    if (pathname.match(/&amp=(y|yes|true|1)/)) {
        pathname = pathname.replace(/&amp=(y|yes|true|1)/, '');
    }
    pathname = pathname.replace(/\?$/, '');
    return pathname;
}
function resultFromChunks(chunks) {
    return ({ next , complete  })=>{
        chunks.forEach(next);
        complete();
        return ()=>{
        };
    };
}
function resultToChunks(result) {
    return new Promise((resolve, reject)=>{
        const chunks = [];
        result({
            next: (chunk)=>{
                chunks.push(chunk);
            },
            error: (error)=>reject(error)
            ,
            complete: ()=>resolve(chunks)
        });
    });
}
const isInternalBlitzMonorepoDevelopment = __dirname.match(/[\\/]packages[\\/]next[\\/]dist[\\/]server$/);
exports.isInternalBlitzMonorepoDevelopment = isInternalBlitzMonorepoDevelopment;
const fixNodeFileTrace = ()=>{
    const path = require('path');
    path.resolve('.blitz.config.compiled.js');
    path.resolve('.next/server/blitz-db.js');
    path.resolve('.next/serverless/blitz-db.js');
};
exports.fixNodeFileTrace = fixNodeFileTrace;
const withFixNodeFileTrace = (fn)=>{
    fixNodeFileTrace();
    return fn;
};
exports.withFixNodeFileTrace = withFixNodeFileTrace;
