// @prisma/sdk/dist/esm/IntrospectionEngine.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
import Debug from "@prisma/debug";
import { BinaryType } from "@prisma/fetch-engine";
import chalk from "chalk";
import { spawn } from "child_process";
import { ErrorArea, RustPanic } from "./panic";
import { resolveBinary } from "./resolveBinary";
import byline from "./utils/byline";
const debugCli = Debug("prisma:introspectionEngine:cli");
const debugRpc = Debug("prisma:introspectionEngine:rpc");
const debugStderr = Debug("prisma:introspectionEngine:stderr");
const debugStdin = Debug("prisma:introspectionEngine:stdin");
class IntrospectionPanic extends Error {
  constructor(message, rustStack, request) {
    super(message);
    this.rustStack = rustStack;
    this.request = request;
  }
}
__name(IntrospectionPanic, "IntrospectionPanic");
class IntrospectionError extends Error {
  constructor(message, code) {
    super(message);
    this.code = code;
  }
}
__name(IntrospectionError, "IntrospectionError");
let messageId = 1;
class IntrospectionEngine {
  constructor({ debug, cwd } = {
    debug: false,
    cwd: process.cwd()
  }) {
    this.listeners = {};
    this.messages = [];
    this.isRunning = false;
    if (debug) {
      Debug.enable("IntrospectionEngine*");
    }
    this.debug = Boolean(debug);
    this.cwd = cwd || process.cwd();
  }
  stop() {
    if (this.child) {
      this.child.kill();
      this.isRunning = false;
    }
  }
  rejectAll(err) {
    Object.entries(this.listeners).map(([id, listener]) => {
      listener(null, err);
      delete this.listeners[id];
    });
  }
  registerCallback(id, callback) {
    this.listeners[id] = callback;
  }
  getDatabaseDescription(schema) {
    return this.runCommand(this.getRPCPayload("getDatabaseDescription", { schema }));
  }
  getDatabaseVersion(schema) {
    return this.runCommand(this.getRPCPayload("getDatabaseVersion", { schema }));
  }
  introspect(schema, force, compositeTypeDepth) {
    this.lastUrl = schema;
    return this.runCommand(this.getRPCPayload("introspect", { schema, force, compositeTypeDepth }));
  }
  debugPanic() {
    return this.runCommand(this.getRPCPayload("debugPanic", void 0));
  }
  listDatabases(schema) {
    this.lastUrl = schema;
    return this.runCommand(this.getRPCPayload("listDatabases", { schema }));
  }
  getDatabaseMetadata(schema) {
    this.lastUrl = schema;
    return this.runCommand(this.getRPCPayload("getDatabaseMetadata", { schema }));
  }
  handleResponse(response) {
    let result;
    try {
      result = JSON.parse(response);
    } catch (e) {
      console.error(`Could not parse introspection engine response: ${response.slice(0, 200)}`);
    }
    if (result) {
      if (result.backtrace) {
        console.log(result);
      }
      if (!result.id) {
        console.error(`Response ${JSON.stringify(result)} doesn't have an id and I can't handle that (yet)`);
      }
      if (!this.listeners[result.id]) {
        console.error(`Got result for unknown id ${result.id}`);
      }
      if (this.listeners[result.id]) {
        this.listeners[result.id](result);
        delete this.listeners[result.id];
      }
    }
  }
  init() {
    if (!this.initPromise) {
      this.initPromise = this.internalInit();
    }
    return this.initPromise;
  }
  internalInit() {
    return new Promise(async (resolve, reject) => {
      try {
        const binaryPath = await resolveBinary(BinaryType.introspectionEngine);
        debugRpc("starting introspection engine with binary: " + binaryPath);
        this.child = spawn(binaryPath, {
          env: process.env,
          cwd: this.cwd,
          stdio: ["pipe", "pipe", "pipe"]
        });
        this.isRunning = true;
        this.child.on("error", (err) => {
          console.error("[introspection-engine] error: %s", err);
          this.child?.kill();
          this.rejectAll(err);
          reject(err);
        });
        this.child.stdin?.on("error", (err) => {
          console.error(err);
          this.child?.kill();
        });
        this.child.on("exit", (code) => {
          this.isRunning = false;
          if (code === 255 && this.lastError && this.lastError.is_panic) {
            const err2 = new RustPanic(this.lastError.message, this.lastError.backtrace, this.lastRequest, ErrorArea.INTROSPECTION_CLI, void 0, this.lastUrl);
            this.rejectAll(err2);
            reject(err2);
            return;
          }
          const messages = this.messages.join("\n");
          let err;
          if (code !== 0 || messages.includes("panicked at")) {
            let errorMessage = chalk.red.bold("Error in introspection engine: ") + messages;
            if (this.lastError && this.lastError.msg === "PANIC") {
              errorMessage = serializePanic(this.lastError);
              err = new IntrospectionPanic(errorMessage, messages, this.lastRequest);
            } else if (messages.includes("panicked at")) {
              err = new IntrospectionPanic(errorMessage, messages, this.lastRequest);
            }
            err = err || new Error(errorMessage);
            this.rejectAll(err);
            reject(err);
          }
        });
        this.child.stdin.on("error", (err) => {
          debugStdin(err);
        });
        byline(this.child.stderr).on("data", (data) => {
          const msg = String(data);
          this.messages.push(msg);
          debugStderr(msg);
          try {
            const json = JSON.parse(msg);
            if (json.backtrace) {
              this.lastError = json;
            }
            if (json.level === "ERRO") {
              this.lastError = json;
            }
          } catch (e) {
            debugCli(e);
          }
        });
        byline(this.child.stdout).on("data", (line) => {
          this.handleResponse(String(line));
        });
        setImmediate(() => {
          resolve();
        });
      } catch (e) {
        this.child?.kill();
        reject(e);
      }
    });
  }
  async runCommand(request) {
    await this.init();
    if (process.env.FORCE_PANIC_INTROSPECTION_ENGINE) {
      request = this.getRPCPayload("debugPanic", void 0);
    }
    if (this.child?.killed) {
      throw new Error(`Can't execute ${JSON.stringify(request)} because introspection engine already exited.`);
    }
    return new Promise((resolve, reject) => {
      this.registerCallback(request.id, (response, err) => {
        if (err) {
          return reject(err);
        }
        if (typeof response.result !== "undefined") {
          resolve(response.result);
        } else {
          if (response.error) {
            this.child?.kill();
            debugRpc(response);
            if (response.error.data?.is_panic) {
              const message = response.error.data?.error?.message ?? response.error.message;
              reject(new RustPanic(message, message, request, ErrorArea.INTROSPECTION_CLI, void 0, this.lastUrl));
            } else if (response.error.data?.message) {
              let message = `${response.error.data.message}
`;
              if (response.error.data?.error_code) {
                message = chalk.redBright(`${response.error.data.error_code}

`) + message;
                reject(new IntrospectionError(message, response.error.data.error_code));
              } else {
                reject(new Error(message));
              }
            } else {
              reject(new Error(`${chalk.redBright("Error in RPC")}
 Request: ${JSON.stringify(request, null, 2)}
Response: ${JSON.stringify(response, null, 2)}
${response.error.message}
`));
            }
          } else {
            reject(new Error(`Got invalid RPC response without .result property: ${JSON.stringify(response)}`));
          }
        }
      });
      if (this.child.stdin.destroyed) {
        throw new Error(`Can't execute ${JSON.stringify(request)} because introspection engine is destroyed.`);
      }
      debugRpc("SENDING RPC CALL", JSON.stringify(request));
      this.child.stdin.write(JSON.stringify(request) + "\n");
      this.lastRequest = request;
    });
  }
  getRPCPayload(method, params) {
    return {
      id: messageId++,
      jsonrpc: "2.0",
      method,
      params: params ? [{ ...params }] : void 0
    };
  }
}
__name(IntrospectionEngine, "IntrospectionEngine");
function serializePanic(log) {
  return `${chalk.red.bold("Error in introspection engine.\nReason: ")}
${log.reason} in ${chalk.underline(`${log.file}:${log.line}:${log.column}`)}

Please create an issue in the ${chalk.bold("prisma")} repo with the error \u{1F64F}:
${chalk.underline("https://github.com/prisma/prisma/issues/new")}
`;
}
__name(serializePanic, "serializePanic");
export {
  IntrospectionEngine,
  IntrospectionError,
  IntrospectionPanic
};
