// @prisma/sdk/dist/esm/logger.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
import chalk from "chalk";
const tags = {
  error: chalk.red("prisma:error"),
  warn: chalk.yellow("prisma:warn"),
  info: chalk.cyan("prisma:info"),
  query: chalk.blue("prisma:query")
};
const should = {
  warn: !process.env.PRISMA_DISABLE_WARNINGS
};
function log(...data) {
  console.log(...data);
}
__name(log, "log");
function warn(message, ...optionalParams) {
  if (should.warn) {
    console.warn(`${tags.warn} ${message}`, ...optionalParams);
  }
}
__name(warn, "warn");
function info(message, ...optionalParams) {
  console.info(`${tags.info} ${message}`, ...optionalParams);
}
__name(info, "info");
function error(message, ...optionalParams) {
  console.error(`${tags.error} ${message}`, ...optionalParams);
}
__name(error, "error");
function query(message, ...optionalParams) {
  console.log(`${tags.query} ${message}`, ...optionalParams);
}
__name(query, "query");
export {
  error,
  info,
  log,
  query,
  should,
  tags,
  warn
};
