// blitz/dist/blitz.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var image = require('next/image');
var link = require('next/link');
var app = require('next/app');
var config = require('next/config');
var dynamic = require('next/dynamic');
var head = require('next/head');
var error = require('next/error');
var document = require('next/document');
var script = require('next/script');
var stdlib = require('next/stdlib');
var stdlibServer = require('next/stdlib-server');
var dataClient = require('next/data-client');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var image__default = /*#__PURE__*/_interopDefault(image);



Object.defineProperty(exports, 'Image', {
  enumerable: true,
  get: function () {
    return image__default['default'];
  }
});
Object.defineProperty(exports, 'ErrorComponent', {
  enumerable: true,
  get: function () {
    return error.ErrorComponent;
  }
});
Object.defineProperty(exports, 'BlitzScript', {
  enumerable: true,
  get: function () {
    return document.BlitzScript;
  }
});
Object.defineProperty(exports, 'Document', {
  enumerable: true,
  get: function () {
    return document.Document;
  }
});
Object.defineProperty(exports, 'DocumentHead', {
  enumerable: true,
  get: function () {
    return document.DocumentHead;
  }
});
Object.defineProperty(exports, 'Html', {
  enumerable: true,
  get: function () {
    return document.Html;
  }
});
Object.defineProperty(exports, 'Main', {
  enumerable: true,
  get: function () {
    return document.Main;
  }
});
Object.defineProperty(exports, 'Script', {
  enumerable: true,
  get: function () {
    return script.Script;
  }
});
Object.keys(link).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return link[k];
    }
  });
});
Object.keys(app).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return app[k];
    }
  });
});
Object.keys(config).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return config[k];
    }
  });
});
Object.keys(dynamic).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return dynamic[k];
    }
  });
});
Object.keys(head).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return head[k];
    }
  });
});
Object.keys(stdlib).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return stdlib[k];
    }
  });
});
Object.keys(stdlibServer).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return stdlibServer[k];
    }
  });
});
Object.keys(dataClient).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return dataClient[k];
    }
  });
});
